### PR TITLE
Added the remaining most important hyperparameters to the SB3 example

### PR DESCRIPTION
**Pull request by stefanfausser**
**List of new SB3 arguments:**

- learning_rate
- n_steps
- batch_size
- ent_coef
- clip_range

### CODE REVIEW

**Comment by stefanfausser** [comment]
I don't understand the comments of the quality checker. How does my proposed changes for stable_baselines3_example.py affect sb3_imitation.py?

**Comment by Ivan-267** [comment]
Thanks for your recent PRs!

Here are some questions I'd like to discuss about this PR:

Adding PPO LSTM introduces a new dependency, SB3 contrib, that won't be installed with Godot RL Agents. 

_I wonder if it would work to include SB3 Contrib LSTM as a separate example script instead?_ 

E.g. I added a CleanRL RNN example with some basic user instructions here, it's separate from the CleanRL PPO script (but CleanRL does use separate files anyway): https://github.com/edbeeching/godot_rl_agents/tree/main/examples/clean_rl_gru_ppo

It would also be great if you get a chance to test this with a small example.

I have made a simple memory test env here: https://github.com/edbeeching/godot_rl_agents_examples/tree/main/examples/TestExamples/SimpleMemoryTest

I tested the CleanRL GRU example on it and included results here: https://github.com/edbeeching/godot_rl_agents/pull/250#issuecomment-3736029559

_About adding CL hyperparams, it can be useful for sharing hyperparams or automating, but I also wonder is it better to keep a lightweight example where users can alter the hyperparams in code? That doesn't exclude the option of potentially having another example that is more integrated for ease of use._

The IDE tells you which hyperparams exist, and a bit about what they do, for whatever algorithm you set (PPO, SAC, etc.). For my use so far, I just change the algorithm and hyperparams from the base example in code. While we could add all hyperparams with some work, if we add only some, we still need to modify the script for others, also the same hyperparam name might mean different things in different algorithms. The CleanRL examples do have cl hyperparameter adjustments but I think include all of them, so both cases can be considered. I welcome your opinion here.

As an idea not directly related to this PR, I have been considering an option to not install any RL framework with Godot RL Agents and have installation instructions with each example in an .md file (e.g. for SB3, for Rllib, etc.), but at this point it's just an idea, currently SB3 is installed with GDRL.

About the style error:
I think the version of the style checker has updated, and that's why it's asking for the SB3 imitation update even though you didn't change anything. I updated the formatting there in this PR: https://github.com/edbeeching/godot_rl_agents/pull/251.

So, if you update your branch to the newest version, hopefully it will be fixed.

**Comment by stefanfausser** [comment]
Thank you @Ivan-267 for your elaborative thoughts. As for your points rised, I feel it would be right to have the most basic hyperparameters as arguments with good default values. Furthermore, when stable_baselines3_example.py is changed, then I agree to change all the sb3-examples (including sb3_imitation.py) alike. This solution would also fit stable_baselines3_hp_tuning.py where tuned learning rates, n_steps and batch_size values are returned. 

Thus, I suggest to reduce this pull request to just adding missing parameters, i.e. the learning rate, the minibatch size and the number of steps, and also add these to sb3_imitation.py. 

Then, in another branch + pull request, I will prepare an example for PPO-LSTM with stable-baselines3 and test it accordingly as you suggested.

If this is fine for you then I will move on.

**Comment by Ivan-267** [comment]
Thanks for the further thoughts, I mostly agree and have some comments to consider.

> Thus, I suggest to reduce this pull request to just adding missing parameters, i.e. the learning rate, the minibatch size and the number of steps, and also add these to sb3_imitation.py.
> Then, in another branch + pull request, I will prepare an example for PPO-LSTM with stable-baselines3 and test it accordingly as you suggested.

I agree about separating these changes.

>  Furthermore, when stable_baselines3_example.py is changed, then I agree to change all the sb3-examples (including sb3_imitation.py) alike. This solution would also fit stable_baselines3_hp_tuning.py where tuned learning rates, n_steps and batch_size values are returned.

The imitation example has multiple possible stages of training, and they could differ in learning rate and other hyperparams, and for some (BC if I remember well), learning rate is adjusted in a different way than the others. Just adding for information, you can still make the common params adjustable if you decide to, I'm just noting the differences from the base SB3 example.

For my use case, adjusting these settings in code was not an issue in any of these scripts as the IDE helps (I also sometimes use calculations such as `batch_size = n_steps * env.num_envs`), and I also sometimes change the algorithm or framework to SB3/SBX which is still relatively simple to do in code. 

So, on one side it might be useful to keep it light, simple to see/understand/modify. But, it's also valid and possible we might add more to it in the future to make these changes easier and sharing hyperparams a bit easier too. To address both sides, if the example script gets a bit too large, we can consider having also a light alternative more similar to the current base example. At least for now, I think we should keep the SB3 example SB3 only.

**Comment by stefanfausser** [comment]
I truly don't want to submit a pull request you are not comfortable with @Ivan-267 and I understand that stable_baselines3_example is just an example that could be changed/adapted by anyone.

On the other hand, I feel that stable_baselines3_example.py is somehow the de-facto standard "example" for godot_rl_agents that many others use. However, I think that not all of them understand how to modify the code to change the most important hyperparameters. And with learning_rate, n_steps and minibatch_size the SB3 example would be very feature complete.

Side note: Well of course the number of hidden layers + of hidden neurons of the multi-layer perceptron, which is 64 + 64, could also be configured (see policy_kwargs argument of PPO) but that is probably a bit out of scope.

And you are right about sb3_imitation. So I will not touch it.

What do you think @Ivan-267: Is adding learning_rate, n_steps and minibatch_size to stable_baselines3_example.py fine for you in this pull request?

If yes, then I would modify the PR accordingly.

**Comment by Ivan-267** [comment]
My opinion on this slightly evolved during the discussion, that's why I had some questions and it helped to discuss it. There was and is nothing wrong about the PR itself, it's just a decision on what we decide to organize/shape these scripts as.

The main/important point was about splitting the LSTM due to adding the requirement for SB3 contrib, even though I understand this also causes some duplication, so even there we might consider a unified script in the future, like you already did in this PR. But for now, perhaps we should split them as agreed so we don't add non-SB3 requirements. 

The parts about IDE changes vs CL changes were just thoughts about my own use case (where a lot of things can be changed during experimenting). For that use case, it wouldn't be much simpler to change some of the params from the cl, but it doesn't mean it's always the case (also considering sharing hyperparams and etc.).

_I agree you should/can add those hyperparams to the example as you proposed, your argument is valid, and it is definitely the most standard entry point for the framework. That's a good start for this PR._ 

I change these hyperparams most frequently with PPO:
`lr, n_steps, batch_size, n_epochs, clip_range, target_kl, ent_coef`

You do not have to include all of them or you can add others if you like, I'm just sharing my perspective on the most commonly changed ones for me. I do sometimes change others and architecture settings, but less frequently, so we don't need to consider those the most common ones to begin with, at least from my experience so far.

I was also thinking whether we should add all of them, and possibly multiple algorithms in the future, as that would be the natural progression to consider (add one, add them all), although I would probably add algorithms one by one after doing some testing, in some cases we can easily add ONNX export support for them (for similar algorithms to the currently supported ones). That's where the change gets slightly more involved, but still worth considering. It's at that point I might consider to have both (a lightweight version and a more no-adjustments-needed one) since the example might start to look more like a tiny-framework on top of SB3, but there's nothing wrong with us focusing on things that help ease of use. We can consider that and leave it for a potential future change. 

For almost all of these use cases, a good .md document can also be helpful. Right now, the SB3 example docs are here: https://github.com/edbeeching/godot_rl_agents/blob/main/docs/ADV_STABLE_BASELINES_3.md.

Without moving any of the current scripts, we do have place in the examples folder itself to add a `readme.md`. Perhaps there we could briefly list the examples and link a doc for each where it exists, at least for scripts that don't contain a readme in the same folder. For the future scripts, we could consider the folder structure (script + simple readme doc). What do you think about that?

**Comment by stefanfausser** [comment]
Interesting: The checks were already started while I updated the title of this PR and the content. 

Now this PR focuses solely on adding the following hyperparameters to SB3 example: learning_rate, n_steps, batch_size, ent_coef, clip_range. The names are kept identical to that of SB3 PPO. Furthermore, I added two sections in ADV_STABLE_BASELINES_3.md that explain those hyperparameters a bit.

Hopefully this PR is now fine. And I checked 'make quality' with the correct version of black this time :)

As agreed on, I will propose a SB3-contrib LSTM-PPO example in another PR later.

**Review by Ivan-267 — APPROVED** [review]
Looks OK, I like the addition to the docs. Thanks.

One thing to note is the change of batch size to 32, I assume this was done to make it more compatible with the default setting of n_steps when there is only a single env (so there are only 32 steps in the buffer with n_steps set to 32).

For a future point, we could consider whether these defaults might be increased a bit or not, or just note somewhere that if learning is failing, one possible reason could be that the n_steps and/or the batch size setting is not large enough. 

But, that won't be true for all envs, the optimal setting is very env specific, it's not easy to propose optimal params for an env without testing, and doing a HP sweep on many envs to get a best default on average could take a long time, so I think it's out of scope for this PR.

**Comment by stefanfausser** [comment]
> One thing to note is the change of batch size to 32, I assume this was done to make it more compatible with the default setting of n_steps when there is only a single env (so there are only 32 steps in the buffer with n_steps set to 32).

This is correct. So, with this default, any number of envs work, even uneven numbers. 

Alternatively, we could instead set both n_steps and (mini-) batch_size to 64 by default. Thinking it over, this might even be a better choice as most PPO-examples, including the original PPO-paper, consider batch_size=64 as a reasonable minimum. However, because of the possibility of uneven number of envs, I feel that the batch_size should never be less than n_steps. In original SB3 PPO, batch_size=64 and n_steps=2048 by default.

What do you think @Ivan-267?

**Comment by Ivan-267** [comment]
Yes, num_envs must be considered when setting the value so it really depends on the env setup, you already mentioned this in the docs which is good. 

SB3 default is 64, yes, but also as you noted the n_steps is 2048, so there's multiple batches in an update with a single env.

In some cases I have increased n_steps significantly from the 32 default, but this does depend on the env, how long the episodes are, the rewards, etc.

I don't have test data at the moment to say which value would work better, so I don't have a decisive preference. I do often try a larger value than 32 so partially I lean toward a larger default, but, I also know there will be shorter/simpler envs where learning might be faster due to more frequent updates with fewer n_steps.

Conclusion:
_For more potential stability or better learning in some envs, let's increase it to 64 as you suggest._ It might help in some envs, others might need more n_steps, and some might train faster with less.

Hyperparams are always subject to adjustments per env, so whatever we set, it's not going to be optimal for all. We can also always revise it in the future as needed.

**Comment by stefanfausser** [comment]
Agreed. Now both n_steps and batch_size is 64 by default. Furthermore, I updated the md doc accordingly.

**Comment by Ivan-267** [comment]
> Agreed. Now both n_steps and batch_size is 64 by default. Furthermore, I updated the md doc accordingly.

Looks good to me. Is it fully ready now?

**Comment by stefanfausser** [comment]
> > Agreed. Now both n_steps and batch_size is 64 by default. Furthermore, I updated the md doc accordingly.
> 
> Looks good to me. Is it fully ready now?

Yes
